src/test/mina_automation/executor.ml

```diff
@@ -0,0 +1,79 @@
+open Integration_test_lib
+open Core_kernel
+open Async
+
+module DockerContext = struct
+  type t =
+    { image : string; workdir : string; volume : string; network : string }
+end
+
+type context =
+  | Dune (* application ran from dune exec command *)
+  | Local (* application ran from _build/default folder*)
+  | Debian (* application installed from mina debian package *)
+  | Docker of DockerContext.t
+  | AutoDetect
+(* application ran inside docker container *)
+
+module Executor = struct
+  type t = { official_name : string; dune_name : string; context : context }
+
+  let of_context ~context ~dune_name ~official_name =
+    { context; dune_name; official_name }
+
+  let run_from_debian t ~(args : string list) ?env () =
+    Util.run_cmd_exn ?env "." t.official_name args
+
+  let run_from_dune t ~(args : string list) ?env () =
+    Util.run_cmd_exn ?env "." "dune" ([ "exec"; t.dune_name; "--" ] @ args)
+
+  let run_from_local t ~(args : string list) ?env () =
+    Util.run_cmd_exn ?env "."
+      (Printf.sprintf "_build/default/%s" t.dune_name)
+      args
+
+  let built_name t = Printf.sprintf "_build/default/%s" t.dune_name
+
+  let path t =
+    match%bind Sys.file_exists (built_name t) with
+    | `Yes ->
+        Deferred.return (built_name t)
+    | _ -> (
+        match%bind Sys.file_exists t.official_name with
+        | `Yes ->
+            Deferred.return t.official_name
+        | _ ->
+            Deferred.return t.dune_name )
+
+  let run t ~(args : string list) ?env () =
+    let open Deferred.Let_syntax in
+    let logger = Logger.create () in
+    match t.context with
+    | AutoDetect -> (
+        match%bind Sys.file_exists (built_name t) with
+        | `Yes ->
+            [%log debug] "running from _build/default folder"
+              ~metadata:[ ("app", `String (built_name t)) ] ;
+            run_from_local t ~args ?env ()
+        | _ -> (
+            match%bind Sys.file_exists t.official_name with
```

Comment:
I wonder if instead of detecting an executable in "official path", i.e. `/usr/local/bin/whatever`, we could instead check `PATH` environment variable.

I.e. something like:

```suggestion
let paths = Option.value_map ~f:(String.split ~on:':') ~default:[] (Sys.getenv "PATH") in
let exists_at_path prefix = match Sys.file_exists (prefix ^ "/" ^ t.official_name) with |`Yes -> Some prefix | _ -> None in
match%bind Deferred.List.find_map ~f:exists_at_path paths with
```

This would enable running this script e.g. from within nix